src/lease/mod.rs

```diff
@@ -65,74 +60,94 @@ use std::sync::Arc;
+pub use keep_alive::{EtcdLeaseKeepAliveRequest, EtcdLeaseKeepAliveResponse};
+pub use revoke::{EtcdLeaseRevokeRequest, EtcdLeaseRevokeResponse};
 
 use crate::lazy::{Lazy, Shutdown};
-use crate::proto::etcdserverpb;
-use crate::proto::etcdserverpb::lease_client::LeaseClient;
-use crate::{Error, Result};
+use crate::protos::rpc;
+use crate::protos::rpc_grpc::LeaseClient;
+use crate::Result;
 
+use grpcio::WriteFlags;
+
+/// grant
 mod grant;
+/// keep alive mod
 mod keep_alive;
+/// revoke mod
 mod revoke;
 
-/// LeaseKeepAliveTunnel is a reusable connection for `Lease Keep Alive` operation.
-/// The underlying gRPC method is Bi-directional streaming.
+/// `LeaseKeepAliveTunnel` is a reusable connection for `Lease Keep Alive` operation.
+/// The underlying `gRPC` method is Bi-directional streaming.
 struct LeaseKeepAliveTunnel {
-    req_sender: UnboundedSender<LeaseKeepAliveRequest>,
-    resp_receiver: Option<UnboundedReceiver<Result<LeaseKeepAliveResponse>>>,
-    shutdown: Option<oneshot::Sender<()>>,
+    /// req_sender
+    #[allow(dead_code)]
+    req_sender: Sender<EtcdLeaseKeepAliveRequest>,
+    /// resp_receiver
+    resp_receiver: Option<Receiver<Result<EtcdLeaseKeepAliveResponse>>>,
+    /// shutdown
+    shutdown: Option<Sender<()>>,
 }
 
 impl LeaseKeepAliveTunnel {
-    fn new(mut client: LeaseClient<Channel>) -> Self {
-        let (req_sender, mut req_receiver) = unbounded_channel::<LeaseKeepAliveRequest>();
-        let (resp_sender, resp_receiver) = unbounded_channel::<Result<LeaseKeepAliveResponse>>();
+    /// new method
+    fn new(client: &LeaseClient) -> Self {
+        let (req_sender, req_receiver) = unbounded::<EtcdLeaseKeepAliveRequest>();
+        let (resp_sender, resp_receiver) = unbounded::<Result<EtcdLeaseKeepAliveResponse>>();
 
-        let (shutdown_tx, shutdown_rx) = oneshot::channel();
+        let (shutdown_tx, shutdown_rx) = unbounded();
+        let shutdown_reponse = shutdown_rx.clone();
+        // monitor inbound lease response and transfer to the receiver
+        let (mut client_req_sender, mut client_resp_receiver) = client
+            .lease_keep_alive()
+            .unwrap_or_else(|e| panic!("Fail to lease_keep_alive, the error is: {}", e));
+        smol::spawn(async move {
+            let mut shutdown_rx = shutdown_rx.into_future().fuse();
+            #[allow(clippy::mut_mut)]
+            while let Ok(req) = req_receiver.recv().await {
+                let lease_keep_alive_request: rpc::LeaseKeepAliveRequest = req.into();
 
-        let request = tonic::Request::new(async_stream::stream! {
-            while let Some(req) = req_receiver.recv().await {
-                let pb: etcdserverpb::LeaseKeepAliveRequest = req.into();
-                yield pb;
+                futures::select! {
+                    res = client_req_sender.send((lease_keep_alive_request, WriteFlags::default())).fuse() => res.unwrap_or_else(|e| panic!("Fail to send request, the error is {}", e)),
```

Comment:
this line is too long, try to break it into multiple lines

Reply:
It is hard to break it into multiple lines, since the reason is variable name and the string is long